### PR TITLE
Drill down to location level

### PR DESCRIPTION
**Pull request by kahummer**
fix: https://github.com/onaio/reveal-frontend/issues/872

### CODE REVIEW

File: src/components/forms/JurisdictionSelect/index.tsx [hunk 1]
```diff
@@ -46,14 +46,39 @@ export interface JurisdictionSelectProps<T = SelectOption> extends AsyncSelectPr
...
+      .then((jurisdictionLocationOptions: JurisdictionOption[]) => {
+        /** Check if payload has no name property then use id instead
+         *  If there is no location return no options
+         */
+        if (!jurisdictionStatus) {
+          setFinalLocation(true);
+          if (
+            jurisdictionLocationOptions.length >= 1 &&
+            !jurisdictionStatus &&
+            !jurisdictionLocationOptions[0].properties.name
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionSelect/index.tsx:67]: Why do we need to check for just the first one in the array?  Whatever we do next we need apply to every item in the array.  And keep in mind that we can have multiple items returned.

File: src/components/forms/JurisdictionSelect/index.tsx [hunk 2]
```diff
@@ -46,14 +46,39 @@ export interface JurisdictionSelectProps<T = SelectOption> extends AsyncSelectPr
 export const promiseOptions = (
   service: OpenSRPService,
   paramsToUse: Dictionary,
-  hierarchy: SelectOption[]
+  hierarchy: SelectOption[],
+  jurisdictionStatus: boolean,
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionSelect/index.tsx:50]: What does this relate to?

File: src/components/forms/JurisdictionSelect/index.tsx [hunk 3]
```diff
@@ -46,14 +46,39 @@ export interface JurisdictionSelectProps<T = SelectOption> extends AsyncSelectPr
...
+        if (!jurisdictionStatus) {
+          setFinalLocation(true);
+          if (
+            jurisdictionLocationOptions.length >= 1 &&
+            !jurisdictionStatus &&
+            !jurisdictionLocationOptions[0].properties.name
+          ) {
+            const optionsWithoutName = jurisdictionLocationOptions.map(item => {
+              return { label: item.id, value: item.id };
+            });
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionSelect/index.tsx:71]: Do we need `optionsWithoutName` and `optionsWithoutName`, can we not change this `return { label: item.id, value: item.id }` to use name if available otherwise use id?  I think we can => idk if the syntax is correct but something like `label: item.name ? item.name : item.id`

File: src/containers/pages/MDAPoint/ClientUpload/index.tsx [hunk 4]
```diff
@@ -70,7 +72,7 @@ export const ClientUpload = (props: any) => {
                   <div style={{ display: 'inline-block', width: '24rem' }}>
                     <Field
                       required={true}
-                      component={JurisdictionSelect}
+                      component={JurisdictionSelectHOF}
```
Comment (moshthepitt) [inline src/containers/pages/MDAPoint/ClientUpload/index.tsx:75]: Why this change?

Comment (kahummer) [reply src/containers/pages/MDAPoint/ClientUpload/index.tsx:75]: @moshthepitt JurisdictionSelectHOF returns the JurisdictionSelect with the flag that allows us to drill down to location level. It is inspired by this https://jaredpalmer.com/formik/docs/api/field#example

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 2 above)
Comment (kahummer) [reply src/components/forms/JurisdictionSelect/index.tsx:50]: Builds the hierarchy label on the text field

File: src/components/forms/OrganizationSelect/index.tsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import reducerRegistry from '@onaio/redux-reducer-registry';
-import React, { useEffect } from 'react';
```
Comment (moshthepitt) [inline src/components/forms/OrganizationSelect/index.tsx:2]: I do not think we need any changes to this component.  Can you please revert these?

File: src/components/forms/SimpleOrgSelect/index.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import { OpenSRPService, URLParams } from '../../../services/opensrp';
+import { Organization } from '../../../store/ducks/opensrp/organizations';
+import { SelectOption } from '../JurisdictionSelect';
+
+/** SimpleOrgSelect props */
+export interface SimpleOrgSelectProps<T = SelectOption> extends AsyncSelectProps<T> {
+  apiEndpoint: string /** the OpenSRP API endpoint */;
+  params: URLParams /** extra URL params to send to OpenSRP */;
+  serviceClass: typeof OpenSRPService /** the OpenSRP service */;
+  promiseOptions: any;
```
Comment (moshthepitt) [inline src/components/forms/SimpleOrgSelect/index.tsx:16]: Lets not use any here.

File: src/components/forms/SimpleOrgSelect/index.tsx [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  apiEndpoint: OPENSRP_ORGANIZATION_ENDPOINT,
+  promiseOptions,
+  serviceClass: OpenSRPService,
+};
+
+const SimpleOrgSelect = (props: SimpleOrgSelectProps & FieldProps) => {
+  const { apiEndpoint, field, form, serviceClass } = props;
+  const service = new serviceClass(apiEndpoint);
+  const params = {
+    location_id: form.values.jurisdictions.id,
```
Comment (moshthepitt) [inline src/components/forms/SimpleOrgSelect/index.tsx:46]: Please pass in the jurisdiction id as a prop and not like this.

File: src/components/forms/SimpleOrgSelect/index.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+/** default props for SimpleOrgSelect */
+export const defaultProps: Partial<SimpleOrgSelectProps> = {
+  apiEndpoint: OPENSRP_ORGANIZATION_ENDPOINT,
+  promiseOptions,
+  serviceClass: OpenSRPService,
+};
+
+const SimpleOrgSelect = (props: SimpleOrgSelectProps & FieldProps) => {
```
Comment (moshthepitt) [inline src/components/forms/SimpleOrgSelect/index.tsx:42]: Add a docstring here

File: src/components/forms/SimpleOrgSelect/tests/index.test.tsx [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  beforeEach(() => {
+    jest.resetAllMocks();
+    jest.clearAllMocks();
+  });
+
+  it('renders without crashing', () => {
+    shallow(<SimpleOrgSelect />);
+    mount(<SimpleOrgSelect {...defaultProps} />);
+  });
+});
```
Comment (moshthepitt) [inline src/components/forms/SimpleOrgSelect/tests/index.test.tsx:16]: Can we write better tests?

File: src/containers/pages/MDAPoint/ClientListView/helpers/serviceHooks.ts [hunk 10]
```diff
@@ -39,7 +39,7 @@ export const postUploadedFile = async (
   params?: string
 ) => {
   const bearer = `Bearer ${getAccessToken(store.getState())}`;
-  await fetch(`${OPENSRP_API_BASE_URL}/upload/?event_name=Child%20Registration${params}`, {
+  await fetch(`${OPENSRP_API_BASE_URL}upload/?event_name=Child%20Registration${params}`, {
```
Comment (moshthepitt) [inline src/containers/pages/MDAPoint/ClientListView/helpers/serviceHooks.ts:42]: 1.  Lets not have magic strings here
2. `Child%20Registration`  should not be hard coded.  It should be a variable/param

File: src/containers/pages/MDAPoint/ClientUpload/tests/__snapshots__/index.test.tsx.snap [hunk 11]
```diff
@@ -77,8 +77,7 @@ exports[`components/ClientUpload Matches snapshot 1`] = `
                             style="display: inline-block; width: 24rem;"
```
Comment (moshthepitt) [inline src/containers/pages/MDAPoint/ClientUpload/tests/__snapshots__/index.test.tsx.snap:77]: We need to make this snapshot MUCH smaller

File: src/components/forms/JurisdictionSelect/index.tsx [hunk 12]
```diff
@@ -46,19 +46,49 @@ export interface JurisdictionSelectProps<T = SelectOption> extends AsyncSelectPr
...
+  new Promise((resolve, reject) => {
     service
-      .list(paramsToUse)
-      .then((e: JurisdictionOption[]) => {
-        const options = e.map(item => {
+      .list({ ...paramsToUse, is_jurisdiction: jurisdictionStatus })
+      .then((jurisdictionLocationOptions: JurisdictionOption[]) => {
+        /** Check if payload has no name property then use id instead
+         *  If there is no location return no options
+         */
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionSelect/index.tsx:61]: Is this comment still valid?

File: src/components/forms/JurisdictionSelect/index.tsx [hunk 13]
```diff
@@ -158,16 +214,21 @@ const JurisdictionSelect = (props: JurisdictionSelectProps & FieldProps) => {
...
       // set the Formik field value
       if (form && field) {
         form.setFieldValue(field.name, '');
       }
     }
   };
-
+  // const clearSelect = () => {
+  //   refs.async.select.state.value = '';
+  // };
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionSelect/index.tsx:227]: Why is this commented out?

File: src/components/forms/LocationSelect/tests/index.test.tsx [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import { mount, shallow } from 'enzyme';
+import React from 'react';
+import { LocationSelect } from '..';
+
+describe('components/forms/LocationSelect', () => {
+  it('renders without crashing', () => {
+    shallow(<LocationSelect />);
+    mount(<LocationSelect />);
+  });
+});
```
Comment (moshthepitt) [inline src/components/forms/LocationSelect/tests/index.test.tsx:10]: Can you write better tests please?  In this case you just need to test that the correct props are passed to the JurisdictionSelect component.

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 13 above)
Comment (moshthepitt) [reply src/components/forms/JurisdictionSelect/index.tsx:227]: Was this meant to implement the reset functionality?

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 12 above)
Comment (kahummer) [reply src/components/forms/JurisdictionSelect/index.tsx:61]: yeah

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 13 above)
Comment (kahummer) [reply src/components/forms/JurisdictionSelect/index.tsx:227]: An attempt but it didn't work

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 12 above)
Comment (moshthepitt) [reply src/components/forms/JurisdictionSelect/index.tsx:68]: Please remove it if it is not valid?

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 13 above)
Comment (moshthepitt) [reply src/components/forms/JurisdictionSelect/index.tsx:234]: I don't think that it is a good idea to have hanging commented out code.  Better to just remove it.

File: src/components/forms/JurisdictionSelect/index.tsx [hunk 15]
```diff
@@ -90,12 +127,22 @@ export const defaultProps: Partial<JurisdictionSelectProps> = {
...
+    labelFieldName,
+    params,
+    serviceClass,
+  } = props;
   const [parentId, setParentId] = useState<string>('');
   const [hierarchy, setHierarchy] = useState<SelectOption[]>([]);
   const [shouldMenuOpen, setShouldMenuOpen] = useState<boolean>(false);
   const [closeMenuOnSelect, setCloseMenuOnSelect] = useState<boolean>(false);
+  const [isJurisdiction, setIsJurisdiction] = useState<boolean>(true);
+  const [lowestLocation, setLowestLocation] = useState<boolean>(false);
```
Comment (moshthepitt) [inline src/components/forms/JurisdictionSelect/index.tsx:145]: `lowestLocation` is never set to `true`.  What is its purpose?  Seems we may not need it

File: src/components/forms/SimpleOrgSelect/tests/index.test.tsx [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+    expect(wrapper.children().props()).toMatchSnapshot('SimpleOrgSelect Select Props ownProps');
+    expect(
+      wrapper
+        .children()
+        .children()
+        .children()
+        .props()
+    ).toMatchSnapshot('SimpleOrgSelect Select Props');
+  });
```
Comment (moshthepitt) [inline src/components/forms/SimpleOrgSelect/tests/index.test.tsx:27]: How does this test show that it "select options correctly"?

File: src/containers/pages/MDAPoint/ClientUpload/tests/index.test.tsx [hunk 17]
```diff
@@ -13,13 +12,16 @@ describe('components/ClientUpload', () => {
     shallow(<ClientUpload />);
   });
 
-  it('Matches snapshot', () => {
+  it('renders props correctly', () => {
```
Comment (moshthepitt) [inline src/containers/pages/MDAPoint/ClientUpload/tests/index.test.tsx:15]: I do not think that you have adequately tested how `ClientUpload` works.

Comment (kahummer) [reply src/containers/pages/MDAPoint/ClientUpload/tests/index.test.tsx:15]: I created an issue to add extensive tests here https://github.com/onaio/reveal-frontend/issues/889.

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 15 above)
Comment (kahummer) [reply src/components/forms/JurisdictionSelect/index.tsx:145]: Am passing it as a param to promiseOptions and it's being set in there as setFinalLocation(true);

File: src/components/forms/SimpleOrgSelect/tests/index.test.tsx (hunk 16 above)
Comment (kahummer) [reply src/components/forms/SimpleOrgSelect/tests/index.test.tsx:27]: my bad just wanted to test the props

File: src/components/forms/JurisdictionSelect/index.tsx (hunk 15 above)
Comment (moshthepitt) [reply src/components/forms/JurisdictionSelect/index.tsx:145]: Cool, thanks for clarifying.
